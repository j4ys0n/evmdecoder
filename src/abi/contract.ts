import { createModuleDebug, TRACE_ENABLED } from '@splunkdlt/debug-logging'
import { cachedAsync, Cache, NoopCache } from '@splunkdlt/cache'
import { EthereumClient } from '../eth/client'
import { getCode } from '../eth/requests'
import { sha3 } from './wasm'
import { Classification } from '../utils/classification'
import { AbiRepository } from './repo'

const { info, debug, trace, warn } = createModuleDebug('abi:contract')

export type Address = string

export interface ContractProxy {
  address: string
  standard?: string
  subType?: string
  target?: string
}

export interface ContractType {
  name: string | null
  subType?: string
  standards?: string[]
  proxies?: ContractProxy[]
  metadata?: boolean
  baseUri?: boolean
  tokenUri?: boolean
  enumeration?: boolean
  receive?: string[]
}

export interface TokenProperties {
  type: string
  name: string
  symbol: string
  decimals?: number
  totalSupply?: number
  baseUri?: string
  items?: string[]
}

export interface MultiSigProperties {
  type: string
  owners: string[]
}

export interface ERC3668Properties {
  type: string
  url: string
}

export interface ContractInfo {
  /** True if the corresponding account is a smart contract, otherwise false */
  isContract: boolean
  /** A unique representation of all function and event signatures of a contract */
  fingerprint?: string
  /** Name of the contract from configured ABIs */
  contractName?: string
  /** Type of contract from signature matching */
  contractType?: ContractType
  /** Token properties (if ERC20, ERC721, ERC777, ERC1155) */
  properties?: TokenProperties | MultiSigProperties | ERC3668Properties
}

/** Lookup function to find matching signature for a hash */
export type SignatureMatcher = (sig: string, address?: Address) => string | undefined

/** Lookup function to find name for a given contract address or fingerprint */
export type ContractNameLookup = (address: Address, fingerprint: string) => string | undefined

/**
 * Find function and event signature hashes in the EVM bytecode and attempt to match them against
 * known hashes from configured ABIs
 */
export function extractFunctionsAndEvents(
  contractCode: string,
  signatureMatcher: SignatureMatcher
): { functions: string[]; events: string[] } | undefined {
  const functionMatches: Set<string> = new Set()
  const eventMatches: Set<string> = new Set()
  const code = Buffer.from(contractCode.slice(2), 'hex')
  for (let i = 0, len = code.length; i < len; i++) {
    const opcode = code[i]
    // Look for PUSH<n> opcodes - their pushData may contain a log topic or function signature hash
    if (opcode >= 0x60 && opcode <= 0x7f) {
      const dataLength = opcode - 0x5f
      if (dataLength === 32 || dataLength === 4) {
        const data = code.slice(i + 1, i + dataLength + 1).toString('hex')
        const match = signatureMatcher(data)
        if (match) {
          ;(dataLength === 4 ? functionMatches : eventMatches).add(match)
        }
      }
      i += dataLength
    }
  }

  const functions = [...functionMatches].sort()
  const events = [...eventMatches].sort()
  return { functions, events }
}

/** Creates a hash of all function and event signatures of a contract used to match ABIs against EVM bytecode */
export function computeContractFingerprint(
  { functions, events }: { functions: string[]; events: string[] } = { functions: [], events: [] }
): string | undefined {
  if (functions.length === 0 && events.length === 0) {
    return
  }
  const fingerprint = sha3(`${functions.join(',')}|${events.join(',')}`)!.slice(2)
  return fingerprint
}

/**
 * Attempts to determine if the given account address is a smart contract and
 * determines some additional information if that's the case
 */
export async function getContractInfo(
  address: Address,
  ethClient: EthereumClient,
  classification: Classification,
  signatureMatcher?: SignatureMatcher,
  contractNameLookup?: ContractNameLookup
): Promise<ContractInfo> {
  debug('Retrieving contract information for address %s', address)
  if (signatureMatcher == null) {
    return { isContract: true }
  }
  const code = await ethClient.request(getCode(address))
  if (code === '0x') {
    return { isContract: false }
  }
  const fnsEvts = extractFunctionsAndEvents(code, (fingerprint: string) => signatureMatcher(fingerprint, address))
  const fingerprint = computeContractFingerprint(fnsEvts)
  if (TRACE_ENABLED) {
    trace('Computed fingerprint %s from contract code %O', fingerprint, fnsEvts)
  }
  const contractName =
    fingerprint != null && contractNameLookup != null ? contractNameLookup(address, fingerprint) : undefined
  const contractType = await classification.classifyContract(address, code)
  const properties = await getContractProperties(classification, contractType, address)
  return {
    isContract: true,
    fingerprint,
    contractName,
    contractType,
    properties
  }
}

async function getContractProperties(
  classification: Classification,
  contractType: ContractType,
  address: string
): Promise<TokenProperties | MultiSigProperties | ERC3668Properties | undefined> {
  const isGnosis = contractType.name != null ? contractType.name.includes('Gnosis') : false
  if (contractType.standards != null || isGnosis) {
    // const target = (Array.isArray(contractType.proxies) && !isGnosis)
    //     ? contractType.proxies[contractType.proxies.length -1].target
    //     : address;
    const target = address
    if (target != null) {
      try {
        const properties = await classification.getContractProperties(target, contractType)
        return properties
      } catch (e) {
        warn(`Couldn't get properties %s (%s)`, target, (e as any).message)
      }
    }
  }
  return undefined
}

export async function contractInfo({
  address,
  ethBatchClient,
  abiRepo,
  contractInfoCache = new NoopCache(),
  classification
}: {
  address: Address
  ethBatchClient: EthereumClient
  abiRepo: AbiRepository
  contractInfoCache: Cache<string, Promise<ContractInfo>>
  classification: Classification
}): Promise<ContractInfo | undefined> {
  if (abiRepo == null) {
    return
  }
  const result = await cachedAsync(address, contractInfoCache, (addr: Address) =>
    getContractInfo(
      addr,
      ethBatchClient,
      classification,
      (sig: string) => abiRepo.getMatchingSignature(sig),
      (address: string, fingerprint: string) =>
        abiRepo.getContractByAddress(address)?.contractName ??
        abiRepo.getContractByFingerprint(fingerprint)?.contractName
    )
  )
  return result
}
