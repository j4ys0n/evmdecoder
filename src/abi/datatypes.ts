import { RuntimeError } from '../utils/error'
import { AbiInput } from './item'
import { getDataSize as wasmGetDataSize, isArrayType as wasmIsArrayType, isValidDataType } from './wasm'

export type AbiType = string

export class DataTypeError extends RuntimeError {}

export function isValidAbiType(typeStr: string): typeStr is AbiType {
  if (!typeStr) {
    return false
  }
  return isValidDataType(typeStr)
}

export function isArrayType(type: AbiType): boolean {
  return wasmIsArrayType(type)
}

export interface DataSize {
  length: number
  /** `false` if a variable-size data type. `length` is the minimum size in this case */
  exact: boolean
}

export function getDataSize(typeStr: AbiType): DataSize {
  return wasmGetDataSize(typeStr) as any
}

export const isTupleArray = (inputDef: AbiInput) => inputDef.type === 'tuple[]'
export const isTuple = (inputDef: AbiInput) => inputDef.type === 'tuple' || isTupleArray(inputDef)

export function encodeInputType(inputDef: AbiInput): string {
  if (isTuple(inputDef)) {
    if (!inputDef.components?.length) {
      throw new DataTypeError('Unable to encode tuple datatype without components')
    }
    const serializedComponentTypes = inputDef.components.map(c => encodeInputType(c)).join(',')
    const serializedType = `(${serializedComponentTypes})`
    return isTupleArray(inputDef) ? `${serializedType}[]` : serializedType
  }
  return inputDef.type
}
