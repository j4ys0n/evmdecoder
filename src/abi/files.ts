import { readdir, readFile, stat, createReadStream } from 'fs-extra'
import { basename, join as joinPath, resolve } from 'path'
import { AbiRepositoryConfig } from '../config'
import { Address } from './contract'
import { createModuleDebug, TRACE_ENABLED } from '@splunkdlt/debug-logging'
import { AbiItemDefinition, AbiItem } from './item'
import { computeContractFingerprint } from './contract'
import { computeSignature } from './signature'
import { createGunzip } from 'zlib'
import BufferList from 'bl'
import { RuntimeError } from '../utils/error'

const { debug, warn, trace } = createModuleDebug('abi:files')

export class AbiError extends RuntimeError {}

interface TruffleBuild {
  contractName: string
  abi: AbiItem[]
  networks?: {
    [id: string]: {
      address: string
    }
  }
}

export interface SignatureFileContents {
  type: 'function' | 'event'
  entries: Array<[string, AbiItemDefinition[]]>
}

export function isAbiItem(obj: any): obj is AbiItem {
  return typeof obj.type === 'string'
}

export function isAbiItemArray(obj: any): obj is AbiItem[] {
  return Array.isArray(obj) && obj.every(isAbiItem)
}

export function isTruffleBuildFile(obj: any): obj is TruffleBuild {
  return typeof obj === 'object' && isAbiItemArray(obj.abi)
}

export function extractDeployedContractAddresses(truffleBuild: TruffleBuild): Address[] | undefined {
  if (truffleBuild.networks) {
    return Object.values(truffleBuild.networks)
      .filter(({ address }) => address != null)
      .map(({ address }) => address.toLowerCase() as Address)
  }
}

export async function* searchAbiFiles(dir: string, config: AbiRepositoryConfig): AsyncIterable<string> {
  debug('Searching for ABI files in %s', dir)
  const dirContents = await readdir(dir).catch(e =>
    Promise.reject(new AbiError(`Failed to load ABIs from directory ${dir}: ${e}`))
  )
  dirContents.sort()
  const subdirs = []
  for (const f of dirContents) {
    const full = joinPath(dir, f)
    const s = await stat(full)
    if (s.isDirectory() && (config.searchRecursive ?? true)) {
      subdirs.push(joinPath(dir, f))
    } else if (s.isFile() && f.endsWith(config.abiFileExtension ?? '.json')) {
      yield full
    }
  }
  for (const sub of subdirs) {
    yield* searchAbiFiles(sub, config)
  }
}

export interface AbiFileContents {
  contractName: string
  contractAddress?: string
  contractFingerprint?: string
  entries: Array<{
    abi: AbiItemDefinition
    sig: string
  }>
  fileName?: string
}

export function computeRelativePath(fileName: string, basePath?: string): string | null {
  if (basePath != null) {
    const resolvedBased = resolve(basePath)
    const resolvedFileName = resolve(fileName)
    if (resolvedFileName.startsWith(resolvedBased)) {
      return resolvedFileName.slice(resolvedBased.length + 1)
    }
  }
  return null
}

export function parseAbiFileContents(
  abiData: any,
  { computeFingerprint, fileName, basePath }: { computeFingerprint: boolean; fileName: string; basePath?: string }
): AbiFileContents {
  let abis: AbiItem[]
  let contractName: string
  let contractAddress: string | undefined
  debug('Parsing contents of ABI file %s', fileName)
  if (isTruffleBuildFile(abiData)) {
    debug('ABI file contains truffle build output')
    abis = abiData.abi
    contractName =
      abiData.contractName ||
      // Fall back to file name without file extension
      basename(fileName).split('.', 1)[0]
    const addresses = extractDeployedContractAddresses(abiData)
    if (addresses != null) {
      if (addresses.length > 1) {
        warn(
          'Found contract %s deployed to multiple (%d) networks, using address %s of first network',
          contractName,
          addresses.length,
          addresses[0]
        )
      }
      contractAddress = addresses[0]
    }
  } else if (isAbiItemArray(abiData)) {
    abis = abiData
    contractName = basename(fileName).split('.', 1)[0]
  } else {
    throw new AbiError(`Invalid contents of ABI file ${fileName}`)
  }

  const entries = abis
    .filter(abi => (abi.type === 'function' || abi.type === 'event') && abi.name != null)
    .map(abi => ({
      abi,
      sig: computeSignature({ name: abi.name!, inputs: abi.inputs ?? [], type: 'function' })
    }))

  let contractFingerprint: string | undefined
  if (computeFingerprint) {
    const functions = entries
      .filter(i => i.abi.type === 'function')
      .map(i => i.sig)
      .sort()
    const events = entries
      .filter(i => i.abi.type === 'event')
      .map(i => i.sig)
      .sort()

    contractFingerprint = computeContractFingerprint({ functions, events })
    if (TRACE_ENABLED) {
      trace(
        'Computed contract fingerprint %s for contract signature %s contents %O',
        contractFingerprint,
        contractName,
        { functions, events }
      )
    } else {
      debug('Computed contract fingerprint %s for contract signature %s', contractFingerprint, contractName)
    }
  }

  return {
    contractName,
    contractAddress,
    contractFingerprint: contractFingerprint,
    entries: entries.map(item => ({
      sig: item.sig,
      abi: {
        name: item.abi.name!,
        type: item.abi.type as 'function' | 'event',
        inputs: item.abi.inputs ?? [],
        contractName,
        contractFingerprint,
        contractAddress
      }
    })),
    fileName: computeRelativePath(fileName, basePath) ?? basename(fileName)
  }
}

export async function loadAbiFile(path: string, config: AbiRepositoryConfig): Promise<AbiFileContents> {
  const contents = await readFile(path, { encoding: 'utf-8' })
  let data
  try {
    data = JSON.parse(contents)
  } catch (e) {
    warn('bad file formatting in abi %s', path)
  }
  if (data === undefined) {
    return Promise.reject()
  }
  return await parseAbiFileContents(data, {
    fileName: path,
    computeFingerprint: config.fingerprintContracts,
    basePath: config.directory
  })
}

export async function readGzipFile(path: string, { encoding = 'utf-8' }: { encoding?: string } = {}): Promise<string> {
  return new Promise((resolve, reject) => {
    const result = new BufferList()
    createReadStream(path)
      .pipe(createGunzip())
      .on('data', chunk => {
        result.append(chunk)
      })
      .on('end', () => {
        resolve(result.toString(encoding))
      })
      .on('error', e => {
        reject(e)
      })
  })
}

export async function loadSignatureFile(path: string): Promise<SignatureFileContents> {
  const contents = await readGzipFile(path)
  return JSON.parse(contents)
}
