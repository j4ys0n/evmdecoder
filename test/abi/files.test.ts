import { loadAbiFile } from '../../src/abi/files'
import { join } from 'path'

describe('loadAbiFile', () => {
  it('loads raw ABI file', async () => {
    const result = await loadAbiFile(join(__dirname, '../abis/BCB.json'), {
      decodeAnonymous: false,
      fingerprintContracts: true,
      directory: join(__dirname, '../abis'),
      requireContractMatch: true,
      reconcileStructShapeFromTuples: false
    })

    expect(result).toMatchSnapshot()
  })

  it('loads truffle build file', async () => {
    await expect(
      loadAbiFile(join(__dirname, '../abis/Airdropper.json'), {
        decodeAnonymous: false,
        fingerprintContracts: true,
        directory: join(__dirname, '..'),
        requireContractMatch: true,
        reconcileStructShapeFromTuples: false
      })
    ).resolves.toMatchSnapshot()
  })

  it('loads EventEmitter.json', async () => {
    await expect(
      loadAbiFile(join(__dirname, '../abis/EventEmitter.json'), {
        decodeAnonymous: false,
        fingerprintContracts: true,
        directory: join(__dirname, '..'),
        requireContractMatch: true,
        reconcileStructShapeFromTuples: false
      })
    ).resolves.toMatchSnapshot()
  })
})
