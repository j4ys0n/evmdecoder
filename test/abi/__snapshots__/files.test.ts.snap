// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadAbiFile loads EventEmitter.json 1`] = `
Object {
  "contractAddress": undefined,
  "contractFingerprint": "a9e90fbc1bf47877d0dd918a13af663585b3290b93c0cd53b0869705ae1137c3",
  "contractName": "EventEmitter",
  "entries": Array [
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "a9e90fbc1bf47877d0dd918a13af663585b3290b93c0cd53b0869705ae1137c3",
        "contractName": "EventEmitter",
        "inputs": Array [
          Object {
            "indexed": false,
            "internalType": "address",
            "name": "_to",
            "type": "address",
          },
          Object {
            "indexed": false,
            "internalType": "uint256",
            "name": "_amount",
            "type": "uint256",
          },
        ],
        "name": "stored",
        "type": "event",
      },
      "sig": "stored(address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "a9e90fbc1bf47877d0dd918a13af663585b3290b93c0cd53b0869705ae1137c3",
        "contractName": "EventEmitter",
        "inputs": Array [],
        "name": "sender",
        "type": "function",
      },
      "sig": "sender()",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "a9e90fbc1bf47877d0dd918a13af663585b3290b93c0cd53b0869705ae1137c3",
        "contractName": "EventEmitter",
        "inputs": Array [
          Object {
            "internalType": "uint256",
            "name": "_amount",
            "type": "uint256",
          },
        ],
        "name": "store",
        "type": "function",
      },
      "sig": "store(uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "a9e90fbc1bf47877d0dd918a13af663585b3290b93c0cd53b0869705ae1137c3",
        "contractName": "EventEmitter",
        "inputs": Array [],
        "name": "value",
        "type": "function",
      },
      "sig": "value()",
    },
  ],
  "fileName": "abis/EventEmitter.json",
}
`;

exports[`loadAbiFile loads raw ABI file 1`] = `
Object {
  "contractAddress": undefined,
  "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
  "contractName": "BCB",
  "entries": Array [
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [],
        "name": "name",
        "type": "function",
      },
      "sig": "name()",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "spender",
            "type": "address",
          },
          Object {
            "name": "value",
            "type": "uint256",
          },
        ],
        "name": "approve",
        "type": "function",
      },
      "sig": "approve(address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [],
        "name": "totalSupply",
        "type": "function",
      },
      "sig": "totalSupply()",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "sender",
            "type": "address",
          },
          Object {
            "name": "recipient",
            "type": "address",
          },
          Object {
            "name": "amount",
            "type": "uint256",
          },
        ],
        "name": "transferFrom",
        "type": "function",
      },
      "sig": "transferFrom(address,address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [],
        "name": "decimals",
        "type": "function",
      },
      "sig": "decimals()",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "spender",
            "type": "address",
          },
          Object {
            "name": "addedValue",
            "type": "uint256",
          },
        ],
        "name": "increaseAllowance",
        "type": "function",
      },
      "sig": "increaseAllowance(address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "account",
            "type": "address",
          },
        ],
        "name": "balanceOf",
        "type": "function",
      },
      "sig": "balanceOf(address)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [],
        "name": "renounceOwnership",
        "type": "function",
      },
      "sig": "renounceOwnership()",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [],
        "name": "owner",
        "type": "function",
      },
      "sig": "owner()",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [],
        "name": "isOwner",
        "type": "function",
      },
      "sig": "isOwner()",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [],
        "name": "symbol",
        "type": "function",
      },
      "sig": "symbol()",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "spender",
            "type": "address",
          },
          Object {
            "name": "subtractedValue",
            "type": "uint256",
          },
        ],
        "name": "decreaseAllowance",
        "type": "function",
      },
      "sig": "decreaseAllowance(address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "recipient",
            "type": "address",
          },
          Object {
            "name": "amount",
            "type": "uint256",
          },
        ],
        "name": "transfer",
        "type": "function",
      },
      "sig": "transfer(address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "owner",
            "type": "address",
          },
          Object {
            "name": "spender",
            "type": "address",
          },
        ],
        "name": "allowance",
        "type": "function",
      },
      "sig": "allowance(address,address)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "newOwner",
            "type": "address",
          },
        ],
        "name": "transferOwnership",
        "type": "function",
      },
      "sig": "transferOwnership(address)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "indexed": true,
            "name": "from",
            "type": "address",
          },
          Object {
            "indexed": true,
            "name": "to",
            "type": "address",
          },
          Object {
            "indexed": false,
            "name": "value",
            "type": "uint256",
          },
          Object {
            "indexed": false,
            "name": "data",
            "type": "bytes",
          },
        ],
        "name": "TransferWithData",
        "type": "event",
      },
      "sig": "TransferWithData(address,address,uint256,bytes)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "indexed": true,
            "name": "previousOwner",
            "type": "address",
          },
          Object {
            "indexed": true,
            "name": "newOwner",
            "type": "address",
          },
        ],
        "name": "OwnershipTransferred",
        "type": "event",
      },
      "sig": "OwnershipTransferred(address,address)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "indexed": true,
            "name": "from",
            "type": "address",
          },
          Object {
            "indexed": true,
            "name": "to",
            "type": "address",
          },
          Object {
            "indexed": false,
            "name": "value",
            "type": "uint256",
          },
        ],
        "name": "Transfer",
        "type": "event",
      },
      "sig": "Transfer(address,address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "indexed": true,
            "name": "owner",
            "type": "address",
          },
          Object {
            "indexed": true,
            "name": "spender",
            "type": "address",
          },
          Object {
            "indexed": false,
            "name": "value",
            "type": "uint256",
          },
        ],
        "name": "Approval",
        "type": "event",
      },
      "sig": "Approval(address,address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "to",
            "type": "address",
          },
          Object {
            "name": "amount",
            "type": "uint256",
          },
        ],
        "name": "mint",
        "type": "function",
      },
      "sig": "mint(address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "from",
            "type": "address",
          },
          Object {
            "name": "value",
            "type": "uint256",
          },
        ],
        "name": "burn",
        "type": "function",
      },
      "sig": "burn(address,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": undefined,
        "contractFingerprint": "30f0d1068a77a3aaa446f680f4aa961c9e981bff9aba4a0962230867d0f3ddf9",
        "contractName": "BCB",
        "inputs": Array [
          Object {
            "name": "to",
            "type": "address",
          },
          Object {
            "name": "value",
            "type": "uint256",
          },
          Object {
            "name": "data",
            "type": "bytes",
          },
        ],
        "name": "transferWithData",
        "type": "function",
      },
      "sig": "transferWithData(address,uint256,bytes)",
    },
  ],
  "fileName": "BCB.json",
}
`;

exports[`loadAbiFile loads truffle build file 1`] = `
Object {
  "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
  "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
  "contractName": "Airdropper",
  "entries": Array [
    Object {
      "abi": Object {
        "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
        "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
        "contractName": "Airdropper",
        "inputs": Array [],
        "name": "renounceOwnership",
        "type": "function",
      },
      "sig": "renounceOwnership()",
    },
    Object {
      "abi": Object {
        "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
        "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
        "contractName": "Airdropper",
        "inputs": Array [],
        "name": "owner",
        "type": "function",
      },
      "sig": "owner()",
    },
    Object {
      "abi": Object {
        "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
        "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
        "contractName": "Airdropper",
        "inputs": Array [],
        "name": "isOwner",
        "type": "function",
      },
      "sig": "isOwner()",
    },
    Object {
      "abi": Object {
        "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
        "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
        "contractName": "Airdropper",
        "inputs": Array [
          Object {
            "name": "newOwner",
            "type": "address",
          },
        ],
        "name": "transferOwnership",
        "type": "function",
      },
      "sig": "transferOwnership(address)",
    },
    Object {
      "abi": Object {
        "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
        "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
        "contractName": "Airdropper",
        "inputs": Array [
          Object {
            "indexed": false,
            "name": "total",
            "type": "uint256",
          },
          Object {
            "indexed": false,
            "name": "tokenAddress",
            "type": "address",
          },
        ],
        "name": "MultiSended",
        "type": "event",
      },
      "sig": "MultiSended(uint256,address)",
    },
    Object {
      "abi": Object {
        "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
        "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
        "contractName": "Airdropper",
        "inputs": Array [
          Object {
            "indexed": true,
            "name": "previousOwner",
            "type": "address",
          },
          Object {
            "indexed": true,
            "name": "newOwner",
            "type": "address",
          },
        ],
        "name": "OwnershipTransferred",
        "type": "event",
      },
      "sig": "OwnershipTransferred(address,address)",
    },
    Object {
      "abi": Object {
        "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
        "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
        "contractName": "Airdropper",
        "inputs": Array [
          Object {
            "name": "value",
            "type": "uint256",
          },
        ],
        "name": "withdraw",
        "type": "function",
      },
      "sig": "withdraw(uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
        "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
        "contractName": "Airdropper",
        "inputs": Array [
          Object {
            "name": "token",
            "type": "address",
          },
          Object {
            "name": "_contributors",
            "type": "address[]",
          },
          Object {
            "name": "_balances",
            "type": "uint256[]",
          },
          Object {
            "name": "fuel",
            "type": "uint256",
          },
          Object {
            "name": "low_limit",
            "type": "uint256",
          },
        ],
        "name": "multiSendToken",
        "type": "function",
      },
      "sig": "multiSendToken(address,address[],uint256[],uint256,uint256)",
    },
    Object {
      "abi": Object {
        "contractAddress": "0xf5e4553d3dcb14523aa8776f67a89bbf7b842327",
        "contractFingerprint": "bc0c577dd0a92316b90fc8f8e37c8e6d0e129ecba3e00a6277bfb0a88b927039",
        "contractName": "Airdropper",
        "inputs": Array [
          Object {
            "name": "badgeToken",
            "type": "address",
          },
          Object {
            "name": "token",
            "type": "address",
          },
          Object {
            "name": "_contributors",
            "type": "address[]",
          },
          Object {
            "name": "_balances",
            "type": "uint256[]",
          },
          Object {
            "name": "_url",
            "type": "string[]",
          },
          Object {
            "name": "fuel",
            "type": "uint256",
          },
          Object {
            "name": "low_limit",
            "type": "uint256",
          },
        ],
        "name": "multiSendBadge",
        "type": "function",
      },
      "sig": "multiSendBadge(address,address,address[],uint256[],string[],uint256,uint256)",
    },
  ],
  "fileName": "abis/Airdropper.json",
}
`;
